File: libdriveless/tests/python/test_world_pose.py
import sys, time
sys.path.append("../../../")
sys.path.append("../../")
import unittest, math
from pydriveless import angle
from pydriveless import WorldPose

class TestWorldPose(unittest.TestCase):

    def test_worldpose_encode_decode(self):
        p1 = WorldPose(angle.new_deg(10.1), angle.new_deg(10.2), 10.3, angle.new_deg(10.4))
        s = str(p1)
        p2 = WorldPose.from_str(s)
        
        self.assertTrue(p1.lat == p2.lat)
        self.assertTrue(p1.lon == p2.lon)
        self.assertEqual(p1.alt, p2.alt)
        self.assertTrue(p1.compass == p2.compass)

    def __compute_diff_percent(self, d1: float, d2: float) -> float:
        return 100*(1 - d1/d2)

    def test_distance_between(self):
        p1 = WorldPose(lat=angle.new_deg(5.566896), lon=angle.new_deg(95.3672), alt=0, compass=angle.new_deg(0))
        p2 = WorldPose(lat=angle.new_deg(5.566607), lon=angle.new_deg(95.370121), alt=0, compass=angle.new_deg(0))
        dist = WorldPose.distance_between(p1, p2)
        
        dist_google = 324.93
        self.assertTrue(self.__compute_diff_percent(dist, dist_google) < 0.5)
        
        p1 = WorldPose(lat=angle.new_deg(5.566896), lon=angle.new_deg(95.3672), alt=0, compass=angle.new_deg(0))
        p2 = WorldPose(lat=angle.new_deg(5.567333), lon=angle.new_deg(95.367886), alt=0, compass=angle.new_deg(0))
        dist_google=89.86
        self.assertTrue(self.__compute_diff_percent(dist, dist_google) < 0.5)

        p1 = WorldPose(lat=angle.new_deg(-29.279371052090724), lon=angle.new_deg(-56.91723210848266), alt=0, compass=angle.new_deg(0))
        p2 = WorldPose(lat=angle.new_deg(-27.92906183475516), lon=angle.new_deg(-49.75414619790477), alt=0, compass=angle.new_deg(0))
        dist = WorldPose.distance_between(p1, p2)
        dist_google=713190
        self.assertTrue(self.__compute_diff_percent(dist, dist_google) < 0.5)

    def test_compute_heading_azimuth(self):
        p1 = WorldPose(lat=angle.new_deg(-29.279371052090724), lon=angle.new_deg(-56.91723210848266), alt=0, compass=angle.new_deg(0))
        p2 = WorldPose(lat=angle.new_deg(-27.92906183475516), lon=angle.new_deg(-49.75414619790477), alt=0, compass=angle.new_deg(0))
        a = WorldPose.compute_heading(p1, p2)
        self.assertAlmostEqual(a.rad(), math.radians(79.61), 2)


if __name__ == "__main__":
    unittest.main()

File: libdriveless/tests/src/test_world_pose.cpp
#include <gtest/gtest.h>
#include "../../include/world_pose.h"
#include "test_utils.h"
#include <cmath>

TEST(StateWorldPose, TestSetGet)
{
    WorldPose p1(angle::rad(1.2), angle::rad(-2.2), 32.1, angle::rad(0.34));

    ASSERT_TRUE(p1.lat() == angle::rad(1.2));
    ASSERT_TRUE(p1.lon() == angle::rad(-2.2));
    ASSERT_FLOAT_EQ(p1.alt(), 32.1);
    ASSERT_TRUE(p1.compass() == angle::rad(0.34));
}

TEST(StateWorldPose, TestOperators)
{
    WorldPose p1(angle::rad(1.2), angle::rad(-2.2), 32.1, angle::rad(0.34));
    WorldPose p2(angle::rad(1.2), angle::rad(-2.2), 32.1, angle::rad(0.34));
    WorldPose p3(angle::rad(2.2), angle::rad(-2.2), 32.1, angle::rad(0.34));

    ASSERT_TRUE(p1 == p2);
    ASSERT_TRUE(p1 != p3);
}

float compute_diff_percent(float d1, float d2)
{
    return 100 * (1 - d1 / d2);
}

TEST(StateWorldPose, TestDistanceBetween)
{
    WorldPose p1(angle::deg(5.566896), angle::deg(95.3672), 0, angle::rad(0));
    WorldPose p2(angle::deg(5.566607), angle::deg(95.370121), 0, angle::rad(0));
    double dist = WorldPose::distanceBetween(p1, p2);

    double dist_google = 324.93;
    ASSERT_TRUE(compute_diff_percent(dist, dist_google) < 0.5);

    WorldPose p3(angle::deg(5.566896), angle::deg(95.3672), 0, angle::rad(0));
    WorldPose p4(angle::deg(5.567333), angle::deg(95.367886), 0, angle::rad(0));
    dist_google = 89.86;
    dist = WorldPose::distanceBetween(p3, p4);
    ASSERT_TRUE(compute_diff_percent(dist, dist_google) < 0.5);

    WorldPose p5(angle::deg(-29.279371052090724), angle::deg(-56.91723210848266), 0, angle::rad(0));
    WorldPose p6(angle::deg(-27.92906183475516), angle::deg(-49.75414619790477), 0, angle::rad(0));
    dist = WorldPose::distanceBetween(p5, p6);
    dist_google = 713190;
    ASSERT_TRUE(compute_diff_percent(dist, dist_google) < 0.5);
}

TEST(StateWorldPose, TestComputeHeading)
{
    WorldPose p1(angle::deg(-29.279371052090724), angle::deg(-56.91723210848266), 0, angle::rad(0));
    WorldPose p2(angle::deg(-27.92906183475516), angle::deg(-49.75414619790477), 0, angle::rad(0));
    angle a = WorldPose::computeHeading(p1, p2);
    ASSERT_TRUE(a == angle::deg(79.61));
}
